Studentized Range Distribution

DominicChm opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->

#### What does this implement/fix?
This adds the Studentized Range distribution, which is commonly used to calculate significance in Tukey's HSD test. 

#### Additional information
<!--Any additional information you think is important.-->
• This distribution uses relatively complex double integrals, so the integrands for the CDF, PDF, and moment functions are implemented in Cython to be used as LowLevelCallables.
• The CDF has a simple (single integral) asymptotic approximation that is switched to at large degrees of freedom.
• To mitigate overflow errors, and to avoid using asymptotic approximations for as long as possible, any term that can be logarithmized without causing errors has been.
• Methods that remain as default (`_entropy`, `_ppf`) do so for a reason. For example, `_entropy` fails to integrate as a triple integral and the default `_ppf` can't (as far as I know) be calculated analytically.
• This distribution has been tested pretty extensively against other implementation, such as those found in R and Statsmodels. Comparison graphs can be found at https://github.com/DominicChm/Studentized-Range-Data
• numpy `ufunc`s are used for vectorization instead of the decorator `@np.vectorize`, because - in testing - it roughly halves execution time.

DominicChm committed:
ENH: Add studentized range distribution

DominicChm committed:
ENH: Integrate studentized-T

DominicChm committed:
TST: fix bugs causing tests to fail.

DominicChm committed:
TST: add studentized_range tests

DominicChm committed:
DOC: initial doc implementation

DominicChm committed:
MAINT: rename studentized_t to studentized_range

DominicChm committed:
MAINT: rename studentized_t to studentized_range

DominicChm committed:
MAINT: move consts inside integral to increase precision

DominicChm committed:
Merge remote-tracking branch 'origin/studentized-t-distribution' into studentized-t-distribution

DominicChm committed:
BUG: Missing underscore

DominicChm committed:
TST: decrease tolerance for table test

DominicChm committed:
Inital commit

DominicChm committed:
DOC: Builds

DominicChm committed:
ENH: Make studentized range CDF use log

DominicChm committed:
ENH: Add custom PDF

DominicChm committed:
MAINT: Fix spelling

DominicChm committed:
ENH: Convert PDF to log, initial moment imp.

DominicChm committed:
TST: Add PDF tests

DominicChm committed:
ENH: Use ufuncs, add debug options

DominicChm committed:
BUG: Fix studentized moment gen fn

DominicChm committed:
ENH: remove PPF, standardize intro+quads

DominicChm committed:
TST: Add moment test

DominicChm committed:
TST: Add moment test

DominicChm committed:
Merge branch 'studentized-testing' into studentized-t-distribution

# Conflicts:
#	scipy/stats/_continuous_distns.py

swallan commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -3990,6 +3990,98 @@ def test_burr_nan_mean_var_9544(self):
+    alpha01 = [[1, 90.03, 227.2, 271.8, 298.0],
+               [3, 8.261, 15.64, 18.22, 19.77],
+               [10, 4.482, 6.875, 7.712, 8.226],
+               [20, 4.024, 5.839, 6.450, 6.823]]
+
+    def test_cdf_against_tables(self):
+        def _test_set_cdf(set, significance):
+            v = set[0]
+            for k_ind in range(3):
+                k = 2 + k_ind * 6
+                q = set[1 + k_ind]
+                res_p = stats.studentized_range.cdf(q, k, v)
+
+                # In assertion, only expect as many sig-figs as we have input
+                assert_allclose(res_p, significance, rtol=1e-4)
+
+        for test_set in self.alpha05:
+            _test_set_cdf(test_set, 0.95)
+
+        for test_set in self.alpha01:
+            _test_set_cdf(test_set, 0.99)
+        return
+
+    @pytest.mark.slow
+    def test_ppf_against_tables(self):
+        def _test_set_ppf(set, significance):
+            v = set[0]
+            for k_ind in range(3):
+                k = 2 + k_ind * 6
+                q = set[1 + k_ind]
+                res_q = stats.studentized_range.ppf(significance, k, v)
+                # Accuracy is somthing to be improved on.
+                assert_allclose(res_q, q, rtol=5e-4)
+
+        for test_set in self.alpha05:
+            _test_set_ppf(test_set, 0.95)
+
+        for test_set in self.alpha01:
+            _test_set_ppf(test_set, 0.99)
+        return
```
```suggestion
 # Critical Q values were selected from this table:
    # https://www.jstor.org/stable/2237810
    # For alpha=.05 and .01, and for each value of v=[1, 3, 10, 20],
    # a Q was picked from each table for k=[2, 8, 14, 20].
    ks = [2, 8, 14, 20]
    vs = [1, 3, 10, 20]

    alpha05 = [17.97, 45.40, 54.33, 59.56,
               4.501, 8.853, 10.35, 11.24,
               3.151, 5.305, 6.028, 6.467,
               2.950, 4.768, 5.357, 5.714]

    alpha01 = [90.03, 227.2, 271.8, 298.0,
               8.261, 15.64, 18.22, 19.77,
               4.482, 6.875, 7.712, 8.226,
               4.024, 5.839, 6.450, 6.823]

    ks_d = np.ravel(list((repeat(ks, 4))))
    vs_d = np.ravel(list(zip(*(repeat(vs, 4)))))
    p95 = [.95] * 16
    p99 = [.99] * 16

    sig_k_v_q = (list(zip(p95, ks_d, vs_d, alpha05))
                 + list(zip(p99, ks_d, vs_d, alpha01)))

    @pytest.mark.parametrize("significance, k, v,q", sig_k_v_q)
    def test_cdf_against_tables(self, significance, k, v, q):
        res_p = stats.studentized_range.cdf(q, k, v)
        # some table values are only accurate to the second decimal place.
        atol = 1e-3 if q < 10 else 1e-2 if q < 100 else 1e-1
        assert_allclose(res_p, significance, atol=atol)

    @pytest.mark.parametrize("significance,k,v,q", sig_k_v_q)
    def test_ppf_against_tables(self, significance, k, v, q):
        res_q = stats.studentized_range.ppf(significance, k, v)
        # some table values are only accurate to the second decimal place.
        atol = 1e-3 if q < 10 else 1e-2 if q < 100 else 1e-1
        assert_allclose(res_q, q, atol=atol)

```

swallan replied:
Parametrized these. I also cross checked the q values against https://www.jstor.org/stable/2237810, which seems to be an original source on the tables.

DominicChm committed:
BUG: Add return statement

swallan committed:
TST: parametrize table comparisons

swallan committed:
TST: missing import

swallan committed:
DOC: add references, clarify comments

swallan commented:
I checked for these things that were changed in gh-12694, they have also been added here.
- [x] update number of continuous dists doc/source/tutorial/stats.rst
- [x] add name to doc/source/tutorial/stats/continuous.rst 
- [x] add tutorial doc/source/tutorial/stats/continuous_studentized_range.rst (added, equations may need to be verified again)
- [x] add to init
- [x] add to _distr_params

swallan committed:
STY: spelling asymptotic

swallan commented on scipy/stats/_continuous_distns.py:
```diff
@@ -8854,6 +8854,139 @@ def _updated_ctor_param(self):
+            # integral takes [a] simpler form," and a single asymptotic
+            # integral is evaluated rather than the standard double integral.
+            # (Lund, R. E., and J. R. Lund, 205)
+            if (v < 2000 and not self._force_fn) or self._force_fn == "regular":
```
@DominicChm is `self._force_fn` temporary or is it intended to be part of the public interface?

Also, is `v < 2000` decided for any particular reason? (In the fortran they switched at 120)

DominicChm replied:
Both `_force_fn` and `_epsabs` were meant to be temporary and used for testing, to enforce either the asymptotic or regular CDF calculation and integral precision respectively. While originally temporary, I'm wondering if it might be useful to leave them in? I can't find any existing distribution classes with instance-wide attributes like this though so it might be best to remove them?

DominicChm replied:
Also, 2000 seems to a safe max before the integral craps out around ~2500-3000.
Edit: Can't remember if this is before it was ln'd. Checking this right now.

swallan replied:
I think that trying to decide on the correct interface would take a lot of time / require reviewer discussion, so I think it would be best to just leave them out.

mdhaber replied:
Yeah, just document the choice.  Maybe mention it in the notes of the docstring.

swallan commented:
Something of note: the docs take a very long time to generate because the standard example docs run the following line

`r = studentized_range.rvs(k, v, size=1000)`

and this command takes 1 + minutes to execute. Should we do anything about that (ie, replacing `%(example)s` with truncated examples?)

mdhaber commented:
Yes, that sounds like a good idea.

DominicChm committed:
DOC: Update tutorial formulas

DominicChm committed:
DOC: Add info to docstring

mdhaber commented:
When you resolve the merge conflicts, also be sure to also add in `_distr_params.py` a set of _invalid_ shape parameters (e.g. one with negative degrees of freedom). Tests will fail otherwise.

swallan committed:
DOC: update note for asymptotic switching

swallan committed:
Merge branch 'studentized-t-distribution' of github.com:DominicChm/scipy into studentized-t-distribution

swallan committed:
DOCS: add a-hat special character

swallan commented on scipy/stats/_continuous_distns.py:
```diff
@@ -8854,6 +8854,149 @@ def _updated_ctor_param(self):
+           Feb. 2021.
+
+
+    %(example)s
```
@mdhaber While looking into adding an example with what was done in gh-13431, I ran into a little issue with the inversion of the cdf. The range of `ks` is based on the support with the specified shapes, but the result is from 0 to infinity, which is troublesome to create a range for. 

```
from scipy.stats import studentized_range as srd
k, v = 3, 10
a, b = srd.support(k, v)
a, b
```
Out[113]: (0, inf)

It may still be possible to generate using just an arbitrarily large number, and then adding infinity onto it. Using np.geomspace would help there be more small-ish samples (There are more `small` q crit values than large, looking at the tables).

I'm not sure if there is a way to evaluate whether this holds up to publishing standards of rvs approximation, but it seems to work experimentally. I'm not sure what the best method to verify this would be -- it doesn't look like the log-likelihood function (from studentized_range._reduce_func...) really works to check data against parameters, results always appear to be infinity. 

So, it might be best to leave it how it is. Here's what I was working with in case you had other thoughts. 

```
import numpy as np
import scipy.stats as stats
from scipy.stats import studentized_range as srd
from scipy.interpolate import interp1d

random_state = np.random.default_rng(123)
size=100
k, v = 3, 10
a, b = srd.support(k, v)

ks = np.geomspace(1, 1200)
ks[-1] = np.inf
# ks = np.arange(a, 1200, dtype=np.float64)
cdf = srd.cdf(ks, k, v)
ppf = interp1d(cdf, ks, kind='next', fill_value='extrapolate')

rvs = ppf(random_state.uniform(size=size))
```

mdhaber replied:
Rather than give away the answer : )
* What is the largest possible random number generated by `rand`?
* Considering how inverse transform sampling works, how can this be used to pick a finite number instead of infinity for the upper end?

swallan replied:
- the range for rand is the open interval of [0, 1]
- Inverse transform sampling uses the inversion of the CDF, the PPF, with uniform input x~[0,1]. So if 1 results in infinity in the arithmatically computer ppf,  picking a value ever so slightly smaller than 1 could get a finite bound. `srd.ppf(1, k, v) -> inf`.
```
In [16]: srd.ppf(1 * (1-np.finfo(float).eps), k, v)
Out[16]: 132.20863262903356
```

mdhaber replied:
Yup! That's what I had in mind. I'd pick a large enough number of points to generate the inverse so that the histogram looks good, but small enough that it takes ~1s or less (if possible) to complete.

swallan replied:
I made histograms with numbers of points in range(100, 1000, 50), and it starts to look good with the sample size around 350, taking ~4s. 

![image](https://user-images.githubusercontent.com/44255917/109097910-22452300-76d5-11eb-9e46-0cb26050617a.png)

mdhaber replied:
Yeah I think we're wasting a lot of resolution at the extreme right tail. `srd.ppf(1 * (1-np.finfo(float).eps), k, v)` is correct if you're going for machine precision, because it's impossible to generate a random number closer to one than that, but in our case, there's a small chance we'll get a number more than, say, 0.999, and even if we do, there will be so few of them that it doesn't matter what their values are; they won't affect the shape of the histogram where it matters. Try `srd.ppf(0.999, k, v)` as the upper end and I bet you can get by with a lot fewer points.

swallan replied:
You're right -- srd.ppf(0.999, k, v) gives  7.41. 

The histogram looks good even with 25 samples, so I think I'll settle it at 50. Then it takes only 400ms.

swallan replied:
[Preview of the docs I uploaded](https://pedantic-kirch-95b16a.netlify.app/scipy.stats.studentized_range%20%E2%80%94%20scipy%20v1.6.0.dev0+052339e%20reference%20guide) so you don't have to regenerate them

swallan committed:
DOC: add example with inverse transform sampling instead of rvs

swallan committed:
DOC: missing line

swallan commented on scipy/stats/_continuous_distns.py:
```diff
@@ -8854,6 +8854,213 @@ def _updated_ctor_param(self):
+    7.41058083802274
+
+    >>> from scipy.interpolate import interp1d
+    >>> xs = np.linspace(a, b, 50)
+    >>> cdf = %(name)s.cdf(xs, k, v)
+    >>> ppf = interp1d(cdf, xs, kind='next', fill_value='extrapolate')
+    >>> random_state = np.random.default_rng(123)
+    >>> r = ppf(random_state.uniform(size=1000))
```
How elaborate should this be explained? above I said what the method was 

 
> Rather than using (``%(name)s.rvs``) to generate random variates, which is
>     very slow for this distribution, we can use inverse transform sampling and
>     interpolation

But that doesn't explain how it actually works. My intuition says that it is a different topic, so an in depth explanantion is not needed other than the terms that could be googled to figure it out.

mdhaber replied:
That's fine not to explain in detail. I'd adjust that to give just a little more, though:
...for this distribution, we can approximate the inverse CDF using an interpolator, and then perform inverse transform sampling with this approximate inverse CDF.

DominicChm committed:
MAINT: Remove debug attributes

swallan committed:
DOC: elaborate on random variate example, remove uneeded notes

note should go in the PR, not the description of the distribution

swallan commented on scipy/stats/_continuous_distns.py:
```diff
@@ -8854,6 +8854,206 @@ def _updated_ctor_param(self):
+    an interpolator, and then perform inverse transform sampling with this
+    approximate inverse CDF.
+
+    Using the shape parameters, obtain the bounds data that will be input to
+    ``cdf`` for later use in interpolation.
+
+    >>> a, b = %(name)s.support(k, v)
+    >>> a, b
+    (0, inf)
+
+    Note that the upper limit for the range for the ``cdf`` input is
+    technically infinite. Inputting 1 to the ``ppf`` gives infinity:
+
+    >>> %(name)s.ppf(1, k, v)
+    inf
+
+    To obtain a finite limit we can input a value that nears 1.
+
+    >>> b = %(name)s.ppf(.999, k, v)
+    >>> b
+    7.41058083802274
```
Use ppf [0, .999] as support is infinite

swallan commented on scipy/stats/_continuous_distns.py:
```diff
@@ -8854,6 +8854,206 @@ def _updated_ctor_param(self):
+    >>> from scipy.interpolate import interp1d
+    >>> xs = np.linspace(a, b, 50)
+    >>> cdf = %(name)s.cdf(xs, k, v)
+    >>> ppf = interp1d(cdf, xs, kind='next', fill_value='extrapolate')
```
create interpolant of inverse cdf

mdhaber replied:
```suggestion
    # Create an interpolant of the inverse CDF
    >>> ppf = interp1d(cdf, xs, fill_value='extrapolate')
```
`kind='next'` was appropriate for a discrete distribution, but not here.

swallan commented on scipy/stats/_continuous_distns.py:
```diff
@@ -8854,6 +8854,206 @@ def _updated_ctor_param(self):
+    >>> cdf = %(name)s.cdf(xs, k, v)
+    >>> ppf = interp1d(cdf, xs, kind='next', fill_value='extrapolate')
+    >>> random_state = np.random.default_rng(123)
+    >>> r = ppf(random_state.uniform(size=1000))
```
perform inverse transform sampling

mdhaber replied:
```suggestion
    # Perform inverse transform sampling using the interpolant
    >>> r = ppf(random_state.uniform(size=1000))
```

swallan reviewed (commented):
make these changes

swallan commented on scipy/stats/_continuous_distns.py:
```diff
@@ -8854,6 +8854,206 @@ def _updated_ctor_param(self):
+
+    See Also
+    --------
+    t: student's t distribution
```
```suggestion
    t: Student's t distribution
```

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -8854,6 +8854,206 @@ def _updated_ctor_param(self):
 
+class studentized_range_gen(rv_continuous):
+    """
+    The generalized studentized range continuous random variable.
```
```suggestion
    The studentized range continuous random variable.
```

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -8854,6 +8854,206 @@ def _updated_ctor_param(self):
+
+    Notes
+    -----
+    When `v` is arbitrarily large, in this implementation greater than or equal
+    to 2000, an asymptotic method is used in calculation of the cumulative
+    distribution function. [4]_
```
When `v` exceeds ___, an asymptotic approximation (infinite degree of freedom) is used to compute the cumulative distribution function.

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -8854,6 +8854,206 @@ def _updated_ctor_param(self):
+    References
+    ----------
+
+    .. [1] "Studentized range distribution",
+           https://en.wikipedia.org/wiki/Studentized_range_distribution
+    .. [2] Batista, Ben Dêivide, et al. "Externally Studentized Normal Midrange
+           Distribution." Ciência e Agrotecnologia, vol. 41, no. 4, 2017, pp.
+           378–389., doi:10.1590/1413-70542017414047716.
+    .. [3] Harter, H. Leon. “Tables of Range and Studentized Range.” The Annals
+           of Mathematical Statistics, vol. 31, no. 4, 1960, pp. 1122–1147.
+           JSTOR, www.jstor.org/stable/2237810. Accessed 18 Feb. 2021.
+    .. [4] Lund, R. E., and J. R. Lund. “Algorithm AS 190: Probabilities and
+           Upper Quantiles for the Studentized Range.” Journal of the Royal
+           Statistical Society. Series C (Applied Statistics), vol. 32, no. 2,
+           1983, pp. 204–210. JSTOR, www.jstor.org/stable/2347300. Accessed 18
+           Feb. 2021.
```
Let's use straight quotation marks throughout. Some citation formats italicize the journal name (not sure which this is); I tend to prefer those. So for [1], for instance, I'd use 
```
"Studentized range distribution". *Wikipedia*.
https://en.wikipedia.org/wiki/Studentized_range_distribution
```

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -8854,6 +8854,206 @@ def _updated_ctor_param(self):
+
+    Examples
+    --------
+    >>> from scipy.stats import %(name)s
```
I would go ahead and substitute in the name here and throughout. I don't care too much, but I imagine that it could come up, whereas I don't think anyone would suggest that you _not_ substitute it in.

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -8854,6 +8854,206 @@ def _updated_ctor_param(self):
+    an interpolator, and then perform inverse transform sampling with this
+    approximate inverse CDF.
+
+    Using the shape parameters, obtain the bounds data that will be input to
+    ``cdf`` for later use in interpolation.
+
+    >>> a, b = %(name)s.support(k, v)
+    >>> a, b
+    (0, inf)
+
+    Note that the upper limit for the range for the ``cdf`` input is
+    technically infinite. Inputting 1 to the ``ppf`` gives infinity:
+
+    >>> %(name)s.ppf(1, k, v)
+    inf
+
+    To obtain a finite limit we can input a value that nears 1.
+
+    >>> b = %(name)s.ppf(.999, k, v)
+    >>> b
+    7.41058083802274
```
```suggestion
    This distribution has an infinite but thin right tail,
    so we focus our attention on the leftmost 99.9%.

    >>> a, b = %(name)s.ppf([0, .999], k, v)
    >>> a, b
    0, 7.41058083802274
```

swallan replied:
Resolved - I used the word 'percent' instead of '%' because the docs refused to generate when writing %.

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -8854,6 +8854,206 @@ def _updated_ctor_param(self):
+    >>> xs = np.linspace(a, b, 50)
+    >>> cdf = %(name)s.cdf(xs, k, v)
+    >>> ppf = interp1d(cdf, xs, kind='next', fill_value='extrapolate')
+    >>> random_state = np.random.default_rng(123)
```
Let's move this below the import above `xs`

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -8854,6 +8854,206 @@ def _updated_ctor_param(self):
+    Calculate a few first moments:
+
+    %(set_vals_stmt)s
+    >>> mean, var, skew, kurt = %(name)s.stats(%(shapes)s, moments='mvsk')
```
First time I run this example code, no problem. Second time I get an error. Did we talk about this? Does that happen on your machine @DominicChm?
```
runfile('C:/Users/matth/Dropbox/Scipy/studentized_range/testing.py', wdir='C:/Users/matth/Dropbox/Scipy/studentized_range', current_namespace=True)
Traceback (most recent call last):

  File "C:\Users\matth\Dropbox\Scipy\studentized_range\testing.py", line 16, in <module>
    x = np.linspace(studentized_range.ppf(0.01, k, v),

  File "C:\Users\matth\scipy\scipy\stats\_distn_infrastructure.py", line 2106, in ppf
    place(output, cond, self._ppf(*goodargs) * scale + loc)

  File "C:\Users\matth\scipy\scipy\stats\_distn_infrastructure.py", line 994, in _ppf
    return self._ppfvec(q, *args)

  File "C:\Users\matth\Anaconda3\envs\scipydev\lib\site-packages\numpy\lib\function_base.py", line 2108, in __call__
    return self._vectorize_call(func=func, args=vargs)

  File "C:\Users\matth\Anaconda3\envs\scipydev\lib\site-packages\numpy\lib\function_base.py", line 2192, in _vectorize_call
    outputs = ufunc(*inputs)

  File "C:\Users\matth\scipy\scipy\stats\_distn_infrastructure.py", line 1769, in _ppf_single
    while self._ppf_to_solve(right, q, *args) < 0.:

  File "C:\Users\matth\scipy\scipy\stats\_distn_infrastructure.py", line 1754, in _ppf_to_solve
    return self.cdf(*(x, )+args)-q

  File "C:\Users\matth\scipy\scipy\stats\_distn_infrastructure.py", line 1929, in cdf
    place(output, cond, self._cdf(*goodargs))

  File "C:\Users\matth\scipy\scipy\stats\_continuous_distns.py", line 9050, in _cdf
    return ufunc(x, k, v).astype("float64")

  File "C:\Users\matth\scipy\scipy\stats\_continuous_distns.py", line 9034, in _single_cdf
    opts=dict(epsabs=1e-11, epsrel=1e-12)

  File "C:\Users\matth\scipy\scipy\integrate\quadpack.py", line 825, in nquad
    return _NQuad(func, ranges, opts, full_output).integrate(*args)

  File "C:\Users\matth\scipy\scipy\integrate\quadpack.py", line 880, in integrate
    **opt)

  File "C:\Users\matth\scipy\scipy\integrate\quadpack.py", line 352, in quad
    points)

  File "C:\Users\matth\scipy\scipy\integrate\quadpack.py", line 465, in _quad
    return _quadpack._qagie(func,bound,infbounds,args,full_output,epsabs,epsrel,limit)

  File "C:\Users\matth\scipy\scipy\integrate\quadpack.py", line 880, in integrate
    **opt)

  File "C:\Users\matth\scipy\scipy\integrate\quadpack.py", line 352, in quad
    points)

  File "C:\Users\matth\scipy\scipy\integrate\quadpack.py", line 465, in _quad
    return _quadpack._qagie(func,bound,infbounds,args,full_output,epsabs,epsrel,limit)

ValueError: invalid callable given
```

DominicChm replied:
Seems like compilation failed in a way that SciPy found the function but couldn't do anything with it. Has this happened since?

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -8854,6 +8854,206 @@ def _updated_ctor_param(self):
+        _a, _b = self._get_support()
+        x = np.atleast_1d(x)
+
+        """The cdf"""
```
```suggestion
```

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -8854,6 +8854,206 @@ def _updated_ctor_param(self):
+    """
+
+    def _argcheck(self, k, v):
+        """Verify args"""
```
```suggestion
```

mdhaber commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -3990,6 +3991,82 @@ def test_burr_nan_mean_var_9544(self):
+    sig_k_v_q = (list(zip(p95, ks_d, vs_d, alpha05))
+                 + list(zip(p99, ks_d, vs_d, alpha01)))
+
+    @pytest.mark.parametrize("significance, k, v,q", sig_k_v_q)
```
```suggestion
    @pytest.mark.parametrize("significance, k, v, q", sig_k_v_q)
```

mdhaber replied:
Actually, don't parameterize this. Just vectorize it (calculate everything in one call to `studentized_range.cdf`)

mdhaber commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -3990,6 +3991,82 @@ def test_burr_nan_mean_var_9544(self):
+        moments = [1, 2, 3, 4]
+
+        for moment in moments:
+            res_act = stats.studentized_range.moment(moment, k, nu)
```
```suggestion
            res_act = stats.studentized_range.moment(moment, k, v)
```
Decide whether to use v or nu throughout.

mdhaber commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -3990,6 +3991,82 @@ def test_burr_nan_mean_var_9544(self):
+            # Setting epsabs to 1e-15 of the PDF + moment quads in studentized
+            # range causes the results to converge and pass default atol.
+            assert_allclose(res_act, res_exp, atol=1e-5)
+
```
Let's test the entropy against trapezoidal rule.

Actually, another thing you could do is generate a data file with a handful of really accurate results with mpmath and test against those values. I think that would strengthen a lot of the tests here because many of them have very loose tolerances. You would also commit the file used to generate the data file.  An example of this sort of thing (using mpmath to generate data, even though it's not used for testing) is in gh-10170.

mdhaber commented on doc/source/tutorial/stats/continuous_studentized_range.rst:
```diff
@@ -0,0 +1,18 @@
+.. math::
+   :nowrap:
+
+    \begin{eqnarray*}
+        f(q;k,\nu) = \frac{\nu^{\nu/2}}{\Gamma(\nu/2)\,2^{\nu/2-1}}\,  \int_0^\infty\int_{-\infty}^\infty  s^{\nu-1}\,e^{-\nu s^{2}/2}  \,sk(k-1) \varphi(z) \varphi(sq+z) [\Phi(sq+z)-\Phi(z)]^{k-2} \,\mathrm{d}z \, \mathrm{d}s\\
+
+        F(q;k,\nu) = \frac{\nu^{\nu/2}}{\Gamma(\nu/2)\,2^{\nu/2-1}}\, \int_0^\infty\int_{-\infty}^\infty s^{\nu-1}\,e^{-\nu s^{2}/2}  k\varphi(z) [\Phi(sq+z)-\Phi(z)]^{k-1} \mathrm{d}z  \, \mathrm{d}s\\
```
@DominicChm @swallan Would the one who didn't write this check carefully against the reference?

Please also include the nu = infinity approximation.

swallan replied:
I didn't write it, so I will check after vectorizing ttest trimmed tonight.

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -8854,6 +8854,206 @@ def _updated_ctor_param(self):
+        K = np.atleast_1d(K)
+
+        def _single_moment(K, k, v):
+            user_data = np.array([K, k, v], float).ctypes.data_as(
+                ctypes.c_void_p)
+
+            llc = LowLevelCallable.from_cython(
+                _stats,
+                '_genstudentized_range_moment',
+                user_data
+            )
+
+            return integrate.nquad(
+                llc,
+                ranges=[(-np.inf, np.inf), (0, np.inf), (_a, _b)],
+                opts=dict(epsabs=1e-11, epsrel=1e-12)
+            )[0]
+
+        return _single_moment(K, k, v)
```
Is the formatting of this stuff copied from somewhere?
It looks like it probably satisfies PEP8 (please run `flake8` on everything and report back, BTW) but I wonder if it could be a little prettier.
For instance, for instance, could
```
 user_data = np.array([K, k, v], float).ctypes.data_as(
                ctypes.c_void_p)
```
be split into two logical lines?

DominicChm replied:
That line is taken from `geninvgauss_gen._cdf`. They managed to fit it all on one line. 
https://github.com/mdhaber/scipy/blob/9ed05e7952493302f76652768d1d08a503544107/scipy/stats/_continuous_distns.py#L3568-L3579

It's possible to split it but that would necessitate (I think) introducing a single-use variable, which I generally try to avoid. 
Ex.
```
data_arr = np.array([K, k, v], float)
user_data = data_arr.ctypes.data_as(ctypes.c_void_p)
```
After checking it, making it a one-liner only goes past the limit by 3 chars, so shortening the name is a possibility, as is just making it a slightly too long one-liner. Opinions?

DominicChm replied:
I've tried to fix this by splitting (for example) `'_genstudentized_range_moment'` into its own variable `cython_symbol`, that both gives some clarification on what the string is and gets many of the split function calls onto one line.

DominicChm committed:
TST: Add test comparing CDF to mpmath

DominicChm committed:
MAINT: Reformat fn calls and var. names in stud. range class

DominicChm committed:
DOC: clarifications and rewording

DominicChm committed:
DOC: add asymptotic approximation

DominicChm committed:
MAINT: replace v with df, correct spacing

swallan committed:
DOC: use names, use leftmost 99.9

swallan committed:
TST: vectorize over parametrize cdf, ppf tests

swallan commented on doc/source/tutorial/stats/continuous_studentized_range.rst:
```diff
@@ -0,0 +1,27 @@
+
+Studentized Range Distribution
+==============================
+This distribution has two shape parameters :math:`\nu>0` and :math:`k>0` and support is :math:`q\geq0`.
```
```suggestion
This distribution has two shape parameters :math:`df>0` and :math:`k>0` and support is :math:`q\geq0`.
```

It seems like we are using `df` internally, should we use `df` here as well?

(I also see some uses of `nu`)

DominicChm replied:
Other distributions (Student's T is a good example) do the same thing, with `v` representing DOF in doc, and some other (usually df) symbol representing it in code.

DominicChm replied:
Other functions use _v_ in their doc to represent DOF and `df` within code, so I did the same. For tests, I think it makes sense to use `v` rather than `df` because we're testing inputs, whose names (I think) should match what's in the doc as closely as possible.

swallan commented on doc/source/tutorial/stats/continuous_studentized_range.rst:
```diff
@@ -0,0 +1,27 @@
+   :nowrap:
+
+    \begin{eqnarray*}
+        f(q;k,\nu) = \frac{\nu^{\nu/2}}{\Gamma(\nu/2)\,2^{\nu/2-1}}\,  \int_0^\infty\int_{-\infty}^\infty  s^{\nu-1}\,e^{-\nu s^{2}/2}  \,sk(k-1) \varphi(z) \varphi(sq+z) [\Phi(sq+z)-\Phi(z)]^{k-2} \,\mathrm{d}z \, \mathrm{d}s\\
```
In the paper by Lund and Lund, they provide a definition of phi and Phi: 
![Screen Shot 2021-03-22 at 5 47 33 PM](https://user-images.githubusercontent.com/44255917/112075946-af459580-8b36-11eb-8b24-1179f9cf79e0.png)

Should we provide this information as well?

mdhaber replied:
You should mention what it is (Gaussian PDF/CDF).

swallan commented on doc/source/tutorial/stats/continuous_studentized_range.rst:
```diff
@@ -0,0 +1,27 @@
+   :nowrap:
+
+    \begin{eqnarray*}
+        f(q;k,\nu) = \frac{\nu^{\nu/2}}{\Gamma(\nu/2)\,2^{\nu/2-1}}\,  \int_0^\infty\int_{-\infty}^\infty  s^{\nu-1}\,e^{-\nu s^{2}/2}  \,sk(k-1) \varphi(z) \varphi(sq+z) [\Phi(sq+z)-\Phi(z)]^{k-2} \,\mathrm{d}z \, \mathrm{d}s\\
```
I checked this against equation (4) from here: https://www.scielo.br/scielo.php?script=sci_arttext&pid=S1413-70542017000400378#e19

And it matches other than the integration being from inf -> 0 and inf -> -inf. Maybe this is possible due to different definitions of `phi`? It matches wikipedia. @DominicChm 

I also wanted to confirm the reason for why _k(k-1)_ is inside the integral is for add'l accuracy, right?

DominicChm replied:
The equations are taken from the Wikipedia. I've checked, and it seems like if you flip the endpoints of both integrals in a double integral, the sign change cancels out. This could just be a different convention.

Regarding the _k(k-1)_ term, I'll move it outside the integral - it's not in there for any particular reason other than that's how I wrote it, and it's probably more clear moved outside.

swallan commented on doc/source/tutorial/stats/continuous_studentized_range.rst:
```diff
@@ -0,0 +1,27 @@
+.. math::
+   :nowrap:
+
+    \begin{eqnarray*}
+        F(q;k,\nu) = k\int_{-\infty}^\infty \varphi(z) [\Phi(q+z)-\Phi(z)]^{k-1} \mathrm{d}z\\
+    \end{eqnarray*}
```
This equation uses  `\Phi(q+z)-\Phi(z)`, which matches wikipedia, but this is what's in[ one of the reference papers](https://www.jstor.org/stable/2347300): 
![Screen Shot 2021-03-22 at 6 53 30 PM](https://user-images.githubusercontent.com/44255917/112080884-f71cea80-8b3f-11eb-8c39-c8adff83dde1.png)

Is the reason for this this our / their Phi is negative the other?

mdhaber replied:
y =Q + z => y - Q = z?

swallan commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -3990,6 +3992,95 @@ def test_burr_nan_mean_var_9544(self):
+    def test_ppf_against_tables(self):
+        res_q = stats.studentized_range.ppf(self.ps, self.ks_d, self.vs_d)
+        # some table values only provide up to the tenths place
+        assert_array_almost_equal(res_q, self.alphas, decimal=1)
```
It was me who made this change with the vectorization, but would it be better to loop over the results that way we can dynamically check what number of decimals to check to, since the reference provides more decimals for smaller numbers?

ie:
```
for res, expect in zip(res_q, self.alphas):
        atol = 1e-3 if expect < 10 else 1e-2 if expect < 100 else 1e-1
        assert_allclose(res, expect, atol=atol)

mdhaber replied:
Just use a constant relative tolerance? I'm guessing you can make that tighter than 1e-1.

DominicChm committed:
DOC: describe phi/Phi in tutorial

DominicChm committed:
DOC: slightly reformat formulas

DominicChm committed:
MAINT: standardize test variable names

DominicChm committed:
Merge remote-tracking branch 'upstream/master' into studentized-t-distribution

# Conflicts:
#	scipy/stats/tests/test_continuous_basic.py
#	scipy/stats/tests/test_distributions.py

mdhaber commented:
I'll update my master.

swallan committed:
ENH: add to invdistcont

swallan committed:
ENH: fix for test_cont_basic.check_named_args

Issue was taking 3 elements, some already numpy arrays, and making 1  numpy array. It raised a warning, meant for jagged array issues. Solution is to make all 3 1d before passing into array.

swallan committed:
TST: omit studentized_range from test_cont_basic.check_fit_args

IntegrationWarning and extreme slowness, warrning skipping these.

swallan committed:
PEP: fix pep8 in tests + sources

swallan committed:
TST: forgotten file arg

swallan committed:
DOC: update number of dists

swallan committed:
TST: skip failing / unreliable studentized range test_fit

Even with supplimental _fitstart method, test_fit struggles with studentized_range: "scipy.integrate.quadpack.IntegrationWarning: The maximum number of subdivisions (50) has been achieved."

swallan committed:
TST: add alpha=.001, make use of itertools product

swallan committed:
TST: test_continuous_basic.py skip `test_rvs_broadcast` and slow tests

swallan committed:
ENH: add _fitstart

swallan committed:
ENH: correct integration range on nu=100,000+

the support returns 0 -> inf, the correct range is -inf -> inf

swallan committed:
TST: test all the way to np.inf

swallan committed:
TST: add invalid log to check_pdf_logpdf_at_endpoints for studentized_range

swallan committed:
DOC: add formula pdf

swallan committed:
DOC: try removing equations

swallan committed:
DOC: add back in PDF

swallan committed:
DOC: add formulas back in

swallan committed:
TST: trying to test linux32 weird np.log errors

swallan committed:
Revert "TST: trying to test linux32 weird np.log errors"

This reverts commit b87f239bbc98e84b3ec086fd6a7468d346085768.

swallan committed:
BUG: define variable for linux32

math.M_SQRT1_2 is not defined on that CI platform

swallan committed:
TST: change product order

swallan committed:
TST: test _fitstart for valid args

swallan committed:
DOC: use full name in docs

swallan committed:
Merge remote-tracking branch 'scipy/master' into studentized-t-distribution

swallan committed:
Merge remote-tracking branch 'scipy/master' into studentized-t-distribution

DominicChm committed:
MAINT: Fix indentation, add constant precomputations

DominicChm committed:
DOC: add some cython fn. docstrings

DominicChm committed:
MAINT: rework argcheck

DominicChm committed:
DOC: Move after_notes

DominicChm committed:
MAINT: Refactor last term position.

DominicChm committed:
MAINT: fix name of asymptotic cython fn

DominicChm committed:
MAINT: clarify _Phi and _phi docstrings

DominicChm committed:
ENH: Vectorize _munp

DominicChm committed:
TST: add moment vectorization test.

DominicChm committed:
TST: parameterize moment test

DominicChm committed:
MAINT: fix typo in test class comment

DominicChm committed:
DOC: clean up equations in tutorial

DominicChm committed:
DOC: update pdf definition in docstring

DominicChm committed:
ENH: replace custom CDF with special.ndtr

DominicChm committed:
ENH: prevent division error warning

DominicChm committed:
DOC: minor changes to cython comments

mdhaber committed:
Merge branch 'master' into studentized-t-distribution

DominicChm committed:
MAINT: remove term from logarithm, remove warning suppression.

DominicChm committed:
STY: code/comment spacing

DominicChm committed:
MAINT: style/old code removal changes for JSON data gen script

DominicChm committed:
DOC: update number of dists

DominicChm committed:
TST: Add PDF mp comparisons

DominicChm committed:
TST: Add moment mp test

DominicChm committed:
Revert "TST: Add moment mp test"

This reverts commit 86073d6e260fb08793e0b19395c331c29a24cab0.

DominicChm committed:
TST: test moment vs mp

DominicChm committed:
MAINT: Update generated data (temp)

swallan committed:
TST: update comment for data

swallan committed:
STY: PEP8 fix in studentizedrange

DominicChm committed:
MAINT: Enable and run all mp reference data

DominicChm committed:
BUG: revert use of cs.ndtr which was innacurate on 32bit

DominicChm committed:
DOC: missing (-) in equations

DominicChm committed:
DOC: Update comments about reason for log

DominicChm committed:
DOC: Fix (-) in docstring

DominicChm committed:
DOC: Add _Phi comment about custom implementation rationale

DominicChm committed:
ENH: Calculate const externally

DominicChm committed:
MAINT: Refactor studentized range to calc constants outside of integrand

DominicChm committed:
BUG: Fix user_data index in moment

swallan committed:
DOC: use single source for q values

swallan committed:
STY: style fixes for prettier code

Not necessarily pep8 issues, but for nicer readability.

swallan committed:
TST: mark slwo studentized range tests

DominicChm committed:
DOC: Fix missing term in tutorial

An s^nu was missing the ^nu portion.

DominicChm committed:
MAINT: Remove usage of support range in some integrals.

DominicChm committed:
MAINT: Remove atleast_1d calls.

atleast_1d was used to ensure outputs of ufunc were np objects, which was needed to call `astype` later. Replace that with a call to np.float64().

DominicChm committed:
DOC: small clarifications in cython/docstring

DominicChm committed:
MAINT: Standardize variable names, PEP8 fixes

DominicChm committed:
MAINT: Rename studentized_range cython functions

Cython funcs had an _gen prefix, which shouldn't have been there. This removes them for clearer naming.

DominicChm committed:
ENH: Add caching mechanism, more test cases to mp generator.

DominicChm committed:
MAINT: Refactor mp PDF

DominicChm committed:
MAINT: Refactor mp moment.

Note: yet to be verified (moments take forever to calculate)

DominicChm committed:
TST: Remove redundant test that compared moment to PDF.

DominicChm committed:
ENH: Generator script overhaul

Adds multiprocessing and case generation through itertools product.

swallan committed:
TST: change variable order

swallan committed:
STY: pep8 fixes in pyx, uneeded citation

swallan committed:
ENH: refactor out duplicate code

swallan committed:
TST: add reference to complement wikipedia

swallan committed:
DOC: add default inf for when v is very large

swallan committed:
DOC: missing ^\nu in PDF on distr. doc.

also, an extra f(q; k, v) statement removed.

swallan committed:
ENH: minor equation refactor + spelling

swallan committed:
MAINT: rename studentized range mpmath test file

swallan committed:
TST: skip studentized_range in test_cont_basic

It takes 4-5x as long as the other slowest distribution

swallan committed:
DOC: clarify large \nu swtich, correct variable name in doc.

mdhaber committed:
Update scipy/stats/_continuous_distns.py

Co-authored-by: Samuel Wallan <44255917+swallan@users.noreply.github.com>

swallan committed:
TST: mark studentized range as xslow in test-cont-basic

swallan committed:
TST: mark studentized range xslow in cases_test_moments

swallan committed:
DOC: update shape constraints in doc for studentized range

swallan committed:
TST: skip to next loop iteration to pass second yield

also, removing the `skip` for studentized range, redundant since it's already xslow.

swallan committed:
DOC: use `x` rather than `x` to make similar to other distributions

swallan committed:
STY: grammar fix

swallan committed:
DOC: simplify wording.

it wasn't rendering correclly before for me locally. This means the same and is more succinct.

mdhaber committed:
Apply suggestions from code review

DominicChm committed:
DOC: Fix self-references to match rename

DominicChm committed:
MAINT: Update data set, add generated data

This expands the number of (fast) tests ran to test a larger sampling of typical cases. MP is now only used to test typical (up to k=20, v=120) cases.

DominicChm committed:
TST: Add R test

Adds a test that compares results for large-v CDF cases against R's ptukey.

DominicChm committed:
MAINT: Remove accessing studentized_range through distributions

Cleans an unneeded access to studentized_range through `distributions`

DominicChm committed:
Merge branch 'studentized-t-distribution' of https://github.com/DominicChm/scipy into studentized-t-distribution

swallan committed:
DOC: adhere to new doctests policy

Co-authored-by: Pamphile ROY <roy.pamphile@gmail.com>
